### ISSUE TITLE
substitute for .bashrc

### ISSUE THREAD

**Issue by pantonox** [microsoft/vscode-remote-release#6618]
Recently, I started using vscode-remote with vscode-cmake-tools. To get it to work, I had to add a "load cmake module" command to the remote .bashrc. Is there a vscode-remote way to run shell commands on the remote machine upon connection such that the remote .bashrc can remain unchanged?

**Comment by tanhakabir** [comment]
Have you tried RemoteCommand? Some details here for what you have to do in the extension to use it: https://github.com/microsoft/vscode-docs/blob/main/remote-release-notes/v1_64.md#enable-remotecommand

**Comment by pantonox** [comment]
I tried the following:
1) Followed your instructions to set remote.SSH.enableRemoteCommand and remote.SSH.useLocalServer
2) Added the following line to ~/.ssh/config
RemoteCommand echo "Hello from vscode remote" && module load tools/binutils && module load devel/CMake/3.18.4-GCCcore-10.2.0 && module load vis/GTK+/3.24.23-GCCcore-10.2.0 && module list && bash
3) Tested ssh config from shell and confirmed module loading on remote machine
4) Connected to host via vscode
    -"Hello from vscode remote" was not printed
    -modules were not loaded
    -cmake-tools not working without loading required modules via .bashrc on remote machine (see also #2467)

**Comment by tanhakabir** [comment]
Could you share your Remote SSH logs of what you see? I'm partially suspicious of `echo "Hello from vscode remote"`

**Comment by pantonox** [comment]
I removed echo "Hello from vscode remote" but this did not change the situation.

The remote machine is located in a university cluster and I do not have the permissions to access remote ssh logs or to modify remote ssh server debugging flags.

My use case is that I have a local machine that tunnels via a cluster frontend to a cluster compute node. First, I start an interactive job on the cluster and then I create a tunnel from the local machine to the cluster compute node as follows:
ssh -L localPortForwarded:clusterComputeNodeName:PortClusterFrontEnd clusterUserName@clusterName -p portComputeNode

On the local machine I can then connect to the clusterComputeNode via the following .ssh/config
Host xxx
	HostName localhost
	User uuu
	Port ####
	IdentityFile /.../.../.ssh/id_rsa
	ForwardX11 yes
	ForwardX11Trusted yes
	ServerAliveInterval 60
	RemoteCommand module load tools/binutils && module load devel/CMake/3.18.4-GCCcore-10.2.0 && module load vis/GTK+/3.24.23-GCCcore-10.2.0 && module list && bash

It seems that the RemoteCommand is not running on the computeNode but on the cluster frontend where loading of modules is failing design (administratively prohibited). If RemoteCommand is not compatible with tunneling of this type it might be best in my case to stay with the .bashrc workflow. What do you think?

**Comment by tanhakabir** [comment]
You're able to access the logs locally. Using these instructions: https://code.visualstudio.com/docs/remote/troubleshooting#_reporting-issues

**Comment by phil-blain** [comment]
If the ["environment modules" package](https://modules.readthedocs.io/en/latest/) providing the `module` command is installed like it usually is, then the `module` command is made available in the environment only for [login shells](https://www.gnu.org/software/bash/manual/html_node/Bash-Startup-Files.html) (see https://modules.readthedocs.io/en/latest/INSTALL.html#configuration). For example on Ubuntu the system package installs `/etc/profile.d/modules.sh` which is sourced for login shells only.

Using a `RemoteCommand` like the one above does not invoke a login shell, so the issue is probably simply that the `module` command is not available. You can check that for yourself by trying from your local machine, once your tunnel to the compute node is established:
~~~
ssh uuu@localhost -p #### -Y -o "RemoteCommand=module load tools/binutils && module load devel/CMake/3.18.4-GCCcore-10.2.0 && module load vis/GTK+/3.24.23-GCCcore-10.2.0 && module list && bash"
~~~

On my end I get: `bash: module: command not found` because the shell where SSH runs the `module` command is not a login shell.

However if I add `source /etc/profile` to the beginning of the `RemoteCommand` then it works, because then I'm forcing the shell to do the same setup as a login shell. So you could try the following:
~~~ssh
RemoteCommand source /etc/profile && module load tools/binutils && module load devel/CMake/3.18.4-GCCcore-10.2.0 && module load vis/GTK+/3.24.23-GCCcore-10.2.0 && module list && bash
~~~

and if you have some environment setup in you own `~/.profile` then simply add it: `RemoteCommand source /etc/profile && source $HOME/.profile && ...`

**Comment by pantonox** [comment]
I tried the suggestion from phil-blain but it did not solve the issue in my case.
Below are the logs requested by tanhakabir. Interestingly the modules are listed in the log but still cmake-tools does not have access.

[18:15:37.001] Log Level: 2
[18:15:37.002] remote-ssh@0.78.0
[18:15:37.002] linux x64
[18:15:37.016] SSH Resolver called for "ssh-remote+7b22686f73744e616d65223a22497269734e6f6465496e746572616374697665227d", attempt 1
[18:15:37.016] "remote.SSH.useLocalServer": true
[18:15:37.016] "remote.SSH.path": undefined
[18:15:37.016] "remote.SSH.configFile": undefined
[18:15:37.017] "remote.SSH.useFlock": true
[18:15:37.017] "remote.SSH.lockfilesInTmp": false
[18:15:37.017] "remote.SSH.localServerDownload": auto
[18:15:37.017] "remote.SSH.remoteServerListenOnSocket": false
[18:15:37.017] "remote.SSH.showLoginTerminal": false
[18:15:37.017] "remote.SSH.defaultExtensions": []
[18:15:37.017] "remote.SSH.loglevel": 2
[18:15:37.017] "remote.SSH.enableDynamicForwarding": true
[18:15:37.017] "remote.SSH.enableRemoteCommand": true
[18:15:37.018] "remote.SSH.serverPickPortsFromRange": {}
[18:15:37.018] "remote.SSH.serverInstallPath": {}
[18:15:37.022] SSH Resolver called for host: IrisNodeInteractive
[18:15:37.022] Setting up SSH remote "IrisNodeInteractive"
[18:15:37.024] Acquiring local install lock: /tmp/vscode-remote-ssh-357951c2-install.lock
[18:15:37.036] Looking for existing server data file at /home/myusername/.config/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-357951c2-dfd34e8260c270da74b5c2d86d61aee4b6d56977-0.78.0/data.json
[18:15:37.036] Using commit id "dfd34e8260c270da74b5c2d86d61aee4b6d56977" and quality "stable" for server
[18:15:37.038] Install and start server if needed
[18:15:37.040] PATH: /home/myusername/miniconda3/envs/hcs/bin:/home/myusername/miniconda3/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
[18:15:37.040] Checking ssh with "ssh -V"
[18:15:37.053] > OpenSSH_8.2p1 Ubuntu-4ubuntu0.4, OpenSSL 1.1.1f  31 Mar 2020

[18:15:37.057] askpass server listening on /run/user/1000/vscode-ssh-askpass-8080f4efc0f511335e4a68cf81c341edc2b0b0d9.sock
[18:15:37.057] Spawning local server with {"serverId":1,"ipcHandlePath":"/run/user/1000/vscode-ssh-askpass-5f1ce3fe10381b1e254a081a647e7343af98dc10.sock","sshCommand":"ssh","sshArgs":["-v","-T","-D","46353","-o","ConnectTimeout=15","IrisNodeInteractive"],"serverDataFolderName":".vscode-server","dataFilePath":"/home/myusername/.config/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-357951c2-dfd34e8260c270da74b5c2d86d61aee4b6d56977-0.78.0/data.json"}
[18:15:37.057] Local server env: {"SSH_AUTH_SOCK":"/run/user/1000/keyring/ssh","SHELL":"/bin/bash","DISPLAY":":1","ELECTRON_RUN_AS_NODE":"1","SSH_ASKPASS":"/home/myusername/.vscode/extensions/ms-vscode-remote.remote-ssh-0.78.0/out/local-server/askpass.sh","VSCODE_SSH_ASKPASS_NODE":"/snap/code/94/usr/share/code/code","VSCODE_SSH_ASKPASS_EXTRA_ARGS":"--ms-enable-electron-run-as-node","VSCODE_SSH_ASKPASS_MAIN":"/home/myusername/.vscode/extensions/ms-vscode-remote.remote-ssh-0.78.0/out/askpass-main.js","VSCODE_SSH_ASKPASS_HANDLE":"/run/user/1000/vscode-ssh-askpass-8080f4efc0f511335e4a68cf81c341edc2b0b0d9.sock"}
[18:15:37.062] Spawned 903854
[18:15:37.160] > local-server-1> Spawned ssh, pid=903863
[18:15:37.166] stderr> OpenSSH_8.2p1 Ubuntu-4ubuntu0.4, OpenSSL 1.1.1f  31 Mar 2020
[18:15:37.201] stderr> debug1: Server host key: ecdsa-sha2-nistp256 SHA256:ZxHBIUSHS1V27iC2oCAQxMu+ua7JKSOvDO3oB1kYCz8
[18:15:37.209] stderr> No Kerberos credentials available (default cache: FILE:/tmp/krb5cc_1000)
[18:15:37.209] stderr> 
[18:15:37.209] stderr> 
[18:15:37.210] stderr> No Kerberos credentials available (default cache: FILE:/tmp/krb5cc_1000)
[18:15:37.210] stderr> 
[18:15:37.210] stderr> 
[18:15:37.486] stderr> Authenticated to localhost ([127.0.0.1]:6000).
[18:15:38.172] > iris-038
[18:15:39.160] stderr> 
[18:15:39.160] stderr> Currently Loaded Modules:
[18:15:39.160] stderr>   1) tools/binutils/2.35
[18:15:39.161] stderr>   2) compiler/GCCcore/10.2.0
[18:15:39.161] stderr>   3) devel/ncurses/6.2-GCCcore-10.2.0
[18:15:39.161] stderr>   4) lib/zlib/1.2.11-GCCcore-10.2.0
[18:15:39.161] stderr>   5) tools/bzip2/1.0.8-GCCcore-10.2.0
[18:15:39.161] stderr>   6) tools/cURL/7.72.0-GCCcore-10.2.0
[18:15:39.161] stderr>   7) tools/libarchive/3.4.3-GCCcore-10.2.0
[18:15:39.161] stderr>   8) devel/CMake/3.18.4-GCCcore-10.2.0
[18:15:39.161] stderr>   9) lib/libffi/3.3-GCCcore-10.2.0
[18:15:39.162] stderr>  10) tools/XZ/5.2.5-GCCcore-10.2.0
[18:15:39.162] stderr>  11) lib/libxml2/2.9.10-GCCcore-10.2.0
[18:15:39.162] stderr>  12) tools/gettext/0.21-GCCcore-10.2.0
[18:15:39.162] stderr>  13) devel/PCRE/8.44-GCCcore-10.2.0
[18:15:39.162] stderr>  14) tools/util-linux/2.36-GCCcore-10.2.0
[18:15:39.162] stderr>  15) vis/GLib/2.66.1-GCCcore-10.2.0
[18:15:39.162] stderr>  16) vis/ATK/2.36.0-GCCcore-10.2.0
[18:15:39.163] stderr>  17) tools/expat/2.2.9-GCCcore-10.2.0
[18:15:39.163] stderr>  18) devel/DBus/1.13.18-GCCcore-10.2.0
[18:15:39.163] stderr>  19) lib/libpng/1.6.37-GCCcore-10.2.0
[18:15:39.163] stderr>  20) vis/freetype/2.10.3-GCCcore-10.2.0
[18:15:39.163] stderr>  21) vis/fontconfig/2.13.92-GCCcore-10.2.0
[18:15:39.163] stderr>  22) devel/xorg-macros/1.19.2-GCCcore-10.2.0
[18:15:39.163] stderr>  23) system/libpciaccess/0.16-GCCcore-10.2.0
[18:15:39.163] stderr>  24) vis/X11/20201008-GCCcore-10.2.0
[18:15:39.164] stderr>  25) vis/at-spi2-core/2.38.0-GCCcore-10.2.0
[18:15:39.164] stderr>  26) vis/at-spi2-atk/2.38.0-GCCcore-10.2.0
[18:15:39.164] stderr>  27) lang/NASM/2.15.05-GCCcore-10.2.0
[18:15:39.164] stderr>  28) lib/libjpeg-turbo/2.0.5-GCCcore-10.2.0
[18:15:39.164] stderr>  29) lib/LibTIFF/4.1.0-GCCcore-10.2.0
[18:15:39.164] stderr>  30) vis/Gdk-Pixbuf/2.40.0-GCCcore-10.2.0
[18:15:39.164] stderr>  31) vis/pixman/0.40.0-GCCcore-10.2.0
[18:15:39.164] stderr>  32) vis/cairo/1.16.0-GCCcore-10.2.0
[18:15:39.165] stderr>  33) lib/ICU/67.1-GCCcore-10.2.0
[18:15:39.165] stderr>  34) vis/HarfBuzz/2.6.7-GCCcore-10.2.0
[18:15:39.165] stderr>  35) lang/FriBidi/1.0.10-GCCcore-10.2.0
[18:15:39.165] stderr>  36) vis/Pango/1.47.0-GCCcore-10.2.0
[18:15:39.165] stderr>  37) tools/gzip/1.10-GCCcore-10.2.0
[18:15:39.165] stderr>  38) lib/lz4/1.9.2-GCCcore-10.2.0
[18:15:39.165] stderr>  39) lib/zstd/1.4.5-GCCcore-10.2.0
[18:15:39.166] stderr>  40) lib/libdrm/2.4.102-GCCcore-10.2.0
[18:15:39.166] stderr>  41) lib/libglvnd/1.3.2-GCCcore-10.2.0
[18:15:39.166] stderr>  42) lib/libunwind/1.4.0-GCCcore-10.2.0
[18:15:39.166] stderr>  43) compiler/LLVM/11.0.0-GCCcore-10.2.0
[18:15:39.166] stderr>  44) vis/Mesa/20.2.1-GCCcore-10.2.0
[18:15:39.166] stderr>  45) lib/libepoxy/1.5.4-GCCcore-10.2.0
[18:15:39.166] stderr>  46) vis/GTK+/3.24.23-GCCcore-10.2.0
[18:15:39.166] stderr> 
[18:15:39.166] stderr>  
[18:15:39.166] stderr> 
[18:15:39.226] > ready: d7cddd57dd5e
[18:15:39.232] > Linux 3.10.0-1127.19.1.el7.x86_64 #1 SMP Tue Aug 25 17:23:54 UTC 2020
[18:15:39.232] Platform: linux
[18:15:39.234] > /bin/bash
[18:15:39.234] Shell: bash
[18:15:39.234] Creating bash subshell inside "bash"
[18:15:39.241] > bash version:  4.2.46(2)-release
[18:15:39.255] > d7cddd57dd5e: running
[18:15:39.321] > Acquiring lock on /home/users/myusername/.vscode-server/bin/dfd34e8260c270da74b5c2d86d61aee4b6d56977/vscode-remote-lock.myusername.dfd34e8260c270da74b5c2d86d61aee4b6d56977
[18:15:39.335] > Found existing installation at /home/users/myusername/.vscode-server/bin/dfd34e8260c270da74b5c2d86d61aee4b6d56977...
> Checking /home/users/myusername/.vscode-server/.dfd34e8260c270da74b5c2d86d61aee4b6d56977.log and /home/users/myusername/.vscode-server/.dfd34e8260c270da74b5c2d86d61aee4b6d56977.pid for a running server
[18:15:39.339] > Looking for server with pid: 27667
[18:15:39.389] > Found running server...
[18:15:39.393] >  
> *
> * Reminder: You may only use this software with Visual Studio family products,
> * as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057)
> *
>  
[18:15:39.437] > Checking server status on port 37319 with wget
[18:15:39.448] > d7cddd57dd5e: start
> SSH_AUTH_SOCK====
> DISPLAY==localhost:10.0==
> webUiAccessToken====
> listeningOn==37319==
> osReleaseId==centos==
> arch==x86_64==
> tmpDir==/run/user/5301==
> platform==linux==
> unpackResult====
> didLocalDownload==0==
> downloadTime====
> installTime====
> extInstallTime====
> serverStartTime====
> connectionToken==111aa1aa-aaa1-1aaa-1a11-aaa111aa111a==
> d7cddd57dd5e: end
[18:15:39.449] Received install output: 
SSH_AUTH_SOCK====
DISPLAY==localhost:10.0==
webUiAccessToken====
listeningOn==37319==
osReleaseId==centos==
arch==x86_64==
tmpDir==/run/user/5301==
platform==linux==
unpackResult====
didLocalDownload==0==
downloadTime====
installTime====
extInstallTime====
serverStartTime====
connectionToken==111aa1aa-aaa1-1aaa-1a11-aaa111aa111a==

[18:15:39.449] Remote server is listening on 37319
[18:15:39.450] Parsed server configuration: {"serverConfiguration":{"remoteListeningOn":{"port":37319},"osReleaseId":"centos","arch":"x86_64","webUiAccessToken":"","sshAuthSock":"","display":"localhost:10.0","tmpDir":"/run/user/5301","platform":"linux","connectionToken":"111aa1aa-aaa1-1aaa-1a11-aaa111aa111a"},"installUnpackCode":""}
[18:15:39.451] Persisting server connection details to /home/myusername/.config/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-357951c2-dfd34e8260c270da74b5c2d86d61aee4b6d56977-0.78.0/data.json
[18:15:39.455] Starting forwarding server. localPort 42445 -> socksPort 46353 -> remotePort 37319
[18:15:39.456] Forwarding server listening on 42445
[18:15:39.456] Waiting for ssh tunnel to be ready
[18:15:39.458] [Forwarding server 42445] Got connection 0
[18:15:39.459] Tunneled 37319 to local port 42445
[18:15:39.459] Resolved "ssh-remote+7b22686f73744e616d65223a22497269734e6f6465496e746572616374697665227d" to "127.0.0.1:42445"
[18:15:39.460] Remote extension host environment: {"DISPLAY":"localhost:10.0"}
[18:15:39.466] ------




[18:15:39.481] [Forwarding server 42445] Got connection 1
[18:15:39.498] [Forwarding server 42445] Got connection 2
[18:15:43.933] [Forwarding server 42445] Got connection 3

**Comment by pantonox** [comment]
My interpretation is that the modules are not exported to the subshell that I created just for keeping the terminal open (see && bash in the .ssh.config argument below)

RemoteCommand source /etc/profile && module load tools/binutils && module load devel/CMake/3.18.4-GCCcore-10.2.0 && module load vis/GTK+/3.24.23-GCCcore-10.2.0 && module list && bash

Question: Is there a way to keep the terminal open without appending "&& bash"?

**Comment by tanhakabir** [comment]
Could you clarify what you mean by keeping the terminal open?

It's unexpected that the modules wouldn't be passed down to the subshells since generally subshells inherit the env from their parent.

Just to verify you did use the same RemoteCommand @phil-blain had suggested?
```
RemoteCommand source /etc/profile && module load tools/binutils && module load devel/CMake/3.18.4-GCCcore-10.2.0 && module load vis/GTK+/3.24.23-GCCcore-10.2.0 && module list && bash
```

What happens if you use a slight modification:
```
RemoteCommand source /etc/profile && module load tools/binutils && module load devel/CMake/3.18.4-GCCcore-10.2.0 && module load vis/GTK+/3.24.23-GCCcore-10.2.0 && module list; bash
```

**Comment by pantonox** [comment]
Dear tanhakabir,
While retesting it seems that the problem solved itself. Probably, because I closed all vscode windows in the meantime and restarted from scratch. 

Overview of the current setting:
1) In my cluster frontend's .bashrc,  all module loading was removed.
2) On the local machine in .ssh.config the following argument is used:
RemoteCommand source /etc/profile && module load tools/binutils && module load devel/CMake/3.18.4-GCCcore-10.2.0 && module load vis/GTK+/3.24.23-GCCcore-10.2.0 && module list && bash
3) On the local machine Remote.SSH: Enable Remote Command is enabled
4) On the remote machine Remote.SSH: Enable Remote Command is enabled as well (not sure if this matters)

Tests that passed:
5) module list on vscode remote terminal now lists the modules loaded via RemoteCommand
6) CMake-Tools now works as expected

PS1: By keeping the terminal open I refer to the statement 
"Only commands that leave the SSH connection in an interactive state will work as expected" from https://github.com/microsoft/vscode-docs/blob/main/remote-release-notes/v1_64.md#enable-remotecommand

PS2: Keeping my Cluster frontend .bashrc tidy is really a game changer for me, because my old projects are in production and I work on new projects requiring a completely different environment.
